File: app/src/main/java/ru/melod1n/vk/concurrent/AsyncCallback.java
package ru.melod1n.vk.concurrent;

import android.app.Activity;

import java.lang.ref.WeakReference;

public abstract class AsyncCallback implements Runnable {

    private WeakReference<Activity> reference;

    public AsyncCallback(Activity activity) {
        reference = new WeakReference<>(activity);
    }

    public abstract void ready() throws Exception;

    public abstract void done();

    public abstract void error(Exception e);

    @Override
    public void run() {
        try {
            ready();
        } catch (Exception e) {
            e.printStackTrace();

            if (reference != null && reference.get() != null) {
                reference.get().runOnUiThread(() -> error(e));
            }

            return;
        }

        if (reference != null && reference.get() != null) {
            reference.get().runOnUiThread(this::done);
        }
    }
}

File: app/src/main/java/ru/melod1n/vk/concurrent/AsyncCallback.kt
package ru.melod1n.vk.concurrent

import android.app.Activity
import java.lang.ref.WeakReference

abstract class AsyncCallback(activity: Activity?) : Runnable {
    private val reference: WeakReference<Activity?>?
    @Throws(Exception::class)
    abstract fun ready()

    abstract fun done()
    abstract fun error(e: Exception?)
    override fun run() {
        try {
            ready()
        } catch (e: Exception) {
            e.printStackTrace()
            if (reference?.get() != null) {
                reference.get()!!.runOnUiThread { error(e) }
            }
            return
        }
        if (reference?.get() != null) {
            reference.get()!!.runOnUiThread { done() }
        }
    }

    init {
        reference = WeakReference(activity)
    }
}